Projects/Shapes/Shapes/Shapes/DynamicArray.h

```diff
@@ -0,0 +1,339 @@
+	int GetCapacity() const
+	{
+		return Capacity;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Reserve(const int NewCapacity)
+	{
+		if (NewCapacity <= Capacity)
+		{
+			return false;
+		}
+
+		T* NewArray = new T[NewCapacity];
+		for (int i = 0; i < Size; i++)
+		{
+			NewArray[i] = Array[i];
+		}
+
+		delete[] Array;
+		Array = NewArray;
+
+		Capacity = NewCapacity;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool ShrinkToFit()
+	{
+		if (Capacity == Size)
+		{
+			return false;
+		}
+
+		T* NewArray = new T[Size];
+		for (int i = 0; i < Size; i++)
+		{
+			NewArray[i] = Array[i];
+		}
+
+		delete[] Array;
+		Array = NewArray;
+
+		Capacity = Size;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	//											MODIFIERS
+	//-------------------------------------------------------------------------------------------
+
+	void Clear()
+	{
+		Size = 0;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void PushBack(const T& Item)
+	{
+		if (Size == Capacity)
+		{
+			if (Capacity == 0)
+			{
+				Reserve(1);
+			}
+			Reserve(Capacity * 2);
+		}
+		Array[Size] = Item;
+		Size++;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void PopBack()
+	{
+		Size--;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Insert(const T& Item, const int Index)
+	{
+		if (Index > Size)
+		{
+			return false;
+		}
+
+		if (Index == Size)
+		{
+			PushBack(Item);
+			Size++;
+			return true;
+		}
+
+		if (Size == Capacity)
+		{
+			Reserve(Capacity * 2);
+		}
+		
+		for (int i = Size; i > Index; i--)
+		{
+			Array[i] = Array[i-1];
+		}
+
+		Array[Index] = Item;
+		Size++;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	bool Erase(const int Index)
+	{
+		if (Index > Size)
+		{
+			return false;
+		}
+
+		if (Index == Size)
+		{
+			PopBack();
+			Size--;
+			return true;
+		}
+
+		for (int i = 0; i < Size - 1; i++)
+		{
+			Array[i] = Array[i + 1];
+		}
+		Size--;
+		return true;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Resize(const int NewSize)
+	{
+		if (Size > Capacity)
+		{
+			Reserve(NewSize);
+		}
+		Size = NewSize;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Resize(const int NewSize, const T& Item)
+	{
+		if (Size <= NewSize)
+		{
+			Size = NewSize;
+		}
+		else
+		{
+			if (NewSize > Capacity)
+			{
+				Reserve(NewSize);
+			}
+			for (int i = Size; i < NewSize; i++)
+			{
+				Array[i] = Item;
+			}
+		}
+	}
```

Comment:
This one is a function overload based on the std::vector function. On std::vector, there is an overload so that you are able to not only resize, but fill any missing slots with an item of your choice. This method is part of the "creative" methods for part 13

Reply:
![ILikeThisAlanShaneLewisGIF](https://user-images.githubusercontent.com/36243539/191057283-66777f80-60dc-44ad-8ce2-5a7a9b42831f.gif)

Reply:
You probably should've put this in the Custom Methods category though